gpflow/optimizers/scipy.py

```diff
@@ -97,13 +102,32 @@ def initial_parameters(cls, variables: Sequence[tf.Variable]) -> tf.Tensor:
 
     @classmethod
     def eval_func(
-        cls, closure: LossClosure, variables: Sequence[tf.Variable], compile: bool = True
+        cls,
+        closure: LossClosure,
+        variables: Sequence[tf.Variable],
+        compile: bool = True,
+        allow_unused_variables: bool = False,
     ) -> Callable[[np.ndarray], Tuple[np.ndarray, np.ndarray]]:
+        first_call = True
```

Comment:
Does this need to be tf.constant(True), to ensure the if statement is revisited?

Reply:
> Does this need to be tf.constant(True), to ensure the if statement is revisited?

Depends... I don't really think it's necessary. If this is compiled, then we will always visit the `if first_call` branch, however all the interesting/expensive stuff inside that branch is pure Python, which won't be compiled into the TF graph, so it doesn't hurt that we're always in that branch...

Reply:
If you're sure autograph doesn't compile the enumerate call then I agree that should be fine.

Reply:
I wrote a small benchmark and tried using a `tf.Variable` for this if-statement - that seems to make optimisation 15-100% slower, so I'm not going to do that.